### DISCUSSION TITLE
Not used for a while - launched today and i get a licence error

### DISCUSSION THREAD

**Discussion by LeighTilley** [atoti/atoti#929] [General]
Hey

Apologies if there has been communication regarding licences I may have missed! :)

I just fired up my personal projects and launched Atoti, after a few months, and I see:

![image](https://github.com/user-attachments/assets/ff1e8c8c-84df-4dfa-8afa-15893bcf992e)

I see: On Demand License Server: AWS_EU_WEST_1_PROD

Is something up with your licence server? Or do I need to refresh something my side?

thx

Leigh
https://tilleytech.com

**Comment by andrewyang94** [comment]
Hey Leigh,

Good to hear back from you! Regarding your issue, this is most likely due to an expiration on your cloud license for the Atoti Community Edition. For context, we allow 100 days for users when using the Atoti Community Edition, upon which users must upgrade to the latest version of atoti to renew their free cloud license. Therefore, I think you'll need to do something like...

```
pip install atoti==0.9.0a0
```

to upgrade to the latest version of Atoti and get you back up and running. Let me know if you face any issues!

Thanks,
Andrew

<img width="1723" alt="image" src="https://github.com/user-attachments/assets/dd7580b0-fc2c-4a92-b670-525dd39b2f3f">

**Reply by LeighTilley** [reply]
OK mate thx for the prompt response and hope all is well with you! :)

Leigh
https://tilleytech.com/

**Comment by LeighTilley** [comment]
Hey hey

Long time no speak and Happy New Year!!! :)

I got bogged down with day job however obtaining some balance again and time for some personal projects! :)

As usual it expired, thus just trying to get it updated. it seems that Atoti 0.9 is local, however on your advice last year I used poetry. Poetry env seems stuck on 0.8.14 despite me trying to update, then using add. I'm rusty / do not use python alot, yet, so each time I come to this it's not immediately obvious to me and yet I bet is super simple!

Do you have advice please? Thanks.

![image](https://github.com/user-attachments/assets/e2a3eaf1-4f1d-454b-93f9-7b1634df779a)


Leigh
https://tilleytech.com

**Reply by andrewyang94** [reply]
Hey Leigh,

Happy to hear from you and Happy New Years as well! In regards to your question, we recently moved from `poetry` to `uv` and so the installation instructions should now be as follows (when at the root of the repo directory):

```
pip install uv
uv sync --python 3.10.11
uv run jupyter-lab
```

Let me know if this works for you!

**Comment by LeighTilley** [comment]
Hey

OK thanks mate. I tried it and got an error regarding the project.toml file,

error: No `project` table found in: `C:\Users\Test\atoti\pyproject.toml`

**Reply by andrewyang94** [reply]
Hmmmm.... It should just work. Can you make sure that you pull the latest from our GitHub repository?

In other words:

```
git checkout main
git pull
```

Then do:

```
pip install uv
uv sync --python 3.10.11
uv run jupyter-lab
```

Let me know if that works for you!

**Comment by LeighTilley** [comment]
Hey hey

had a chance to try just now as WFH :)

![image](https://github.com/user-attachments/assets/7defd7c2-d5d5-4f33-a725-2938dc0762c9)

**Reply by andrewyang94** [reply]
I think you may have to do the following:

```
git checkout main
git pull
git reset --hard origin/main
```

to get to the latest commit of the `main` branch. Note that doing this will overwrite your local changes with the latest files from `main`, but I think this is exactly what you need to fix your problem.

This is because I see you're getting the error about your local changes possibly getting overwritten by merge when doing a `git pull`, which is preventing you from pulling the latest files. Hence, why `uv` is complaining about not having a proper `pyproject.toml`, since you don't have the latest `pyproject.toml`.

Try the above and then do:

```
pip install uv
uv sync --python 3.10.11
uv run jupyter-lab
```

I think this should unblock you, but let me know!

**Comment by LeighTilley** [comment]
Hey hey

Good call re --hard reset (i've occasionally had to use that at work with GitLab haha).

So uv looks much cleaner and that all worked. Now:

![image](https://github.com/user-attachments/assets/77f76404-df5a-4078-adff-f3d4cc128527)

**Reply by andrewyang94** [reply]
Perfect, I think what I'm seeing from your screenshot are some errors from the syntax shift from `0.8.X` to `0.9.X` but that could also be because your `tutorial.ipynb` file is from 5 months ago haha. If you have any further issues, don't hesitate to let me know — I'll be here!

Cheers 🍻 
Andrew

**Comment by LeighTilley** [comment]
Hey

OK cool. I deleted my tutorial folder and did

python -m atoti.copy_tutorial tutorial

and it copies down a file that has a date of 12 Aug 2024. So is this the old one then? As this command is still up on the Atoti installation page.

Sorry if im asking simple questions; i only return to it now and then as day job been too busy.. :). And I run it purely to then write test stuff against. :) I can't remember if I explained it, when it is furthher along the line I will.

**Reply by andrewyang94** [reply]
> and it copies down a file that has a date of 12 Aug 2024. So is this the old one then? As this command is still up on the Atoti installation page.

Hmm let me look into this. If this is copying the old tutorial then that's something to bubble up to the R&D folks. Thanks for letting me know!

**Reply by andrewyang94** [reply]
Checking back in, the resulting notebook for `python -m atoti.copy_tutorial tutorial` works for me for the latest version `0.9.3`. My suspicion is that you may have previously used the base python on your local machine when executing `python -m atoti.copy_tutorial tutorial`, which stayed the same version as last time (probably `0.8.X`). 

Instead, since we're using `uv`, you'll have to run that command from within the virtual environment created by `uv` by doing `uv run python -m atoti.copy_tutorial tutorial` and then doing a `uv run jupyter-lab`. I think this should work for you moving forward but let me know otherwise!

As a comparison, you can probably see a difference when you do:

```
pip list | grep atoti
```

vs.

```
uv pip list | grep atoti
```

> **Note:** If you're using Windows, you may have to do `findstr atoti` instead.

**Comment by LeighTilley** [comment]
Hey hey mate

Thanks will try later when at home. I just need to spend a bit more time to learn the Python stuff properly as lately all I work in is C# :)

**Comment by LeighTilley** [comment]
Hey mate

Thx for pointers. All updated and working / see deprecated warning on stuff like session.port (which i use to get port for my custom stuff)

![image](https://github.com/user-attachments/assets/376f52da-360d-4ae2-8a6d-54ea96298798)


thx

Leigh
https://tilleytech.com

**Reply by andrewyang94** [reply]
FYI, to fix this deprecation warning, you might want to try something like this:

```
import atoti as tt
from urllib.parse import urlparse

session = tt.Session.start(
    tt.SessionConfig(
        user_content_storage="./content",
        port=9092
    )
)

url = session.url
port = urlparse(url).port
```

This was a change going from Atoti Python API 0.8 to 0.9, you can find more of these types of changes outlined in https://www.atoti.io/articles/upgrading-to-atoti-python-api-0-9-0/!

**Reply by LeighTilley** [reply]
Hey

Thanks and no worries. I wasn't displaying it as a complaint or moan, it was to highlight that it was a newer version thus some old calls are deprecated. :)

And then also a screenshot using my own custom C# DLL which speaks to remote cubes which I made to use with a project I'm slowly working on.

**Reply by andrewyang94** [reply]
Haha sounds good, well if you have any other questions/concerns/comments, you know where to find me :) — good luck!

**Comment by LeighTilley** [comment]
Hey 

Yes getting data in my initial tester area (the real DLL use will be in Unity :) ):

![image](https://github.com/user-attachments/assets/84c8bb2c-ff4c-4cd1-b773-83e78216c99f)

**Comment by LeighTilley** [comment]
Hey mate

I hope all is well! :)

I got a chance to fire up stuff, and where I need to have a lean codebase in Unity, and it'll run on ARM Android processors, it seems my own custom C# MDX code will not work (as it uses MS ADOMDNet DLL). So I'll need to port it to Java. So i found a very old oLAP4J project, although this stuff expects what it calls a BackEndFlavor, derived from datasourceInfo which seems to be null coming back from your demo Atoti.

![image](https://github.com/user-attachments/assets/4c765b45-cddf-4af9-9c9c-26bc56b1f9f2)

Do you know why? Or can you recommend any other Java kit? Or I guess I could modify it and rebuild and/or look for alternatives.

Thanks

Leigh
https://tilleytech.com

**Reply by andrewyang94** [reply]
Hey Leigh,

Can you provide some more context on what you're trying to achieve here? Based on what I see, it looks like you're trying to connect to the Atoti database and pull some data, right? If so, I think what you may be looking for is making a request to our REST API that should return/respond with the data that you want. You can check out the following documentation for this: https://docs.activeviam.com/products/atoti/server/latest/docs/rest-api/database_rest_api/.

> **Note:** The Atoti Server is bundled with Atoti Python API, so even though the docs say Atoti Server, the information therein should still be applicable even if you're using Atoti Python API to spin up your Atoti application.

Or since you're saying you're porting custom code to Java, it may be helpful for you to try out our Atoti Java API as well: https://github.com/activeviam/atoti-spring-boot-template

Of course, if I'm misunderstanding your use case here, let me know and I can try to re-asses my answer or loop someone else in who may have more information for you. Let me know!

Thanks,
Andrew

**Comment by LeighTilley** [comment]
Hey hey

So my custom project is in Unity games engine for an AR glasses app to visualise data. I began it ages ago and then it's been on hold as day job became too busy. So I came back to it, and then began simple data testing I realised that where the target runtime platform is Android and ARM processors the Microsoft stack I'd used, as I wrote my C# codebase for performing MDX querying, will not be built in. So it performs part of my code and then the underlying MS native cube connections stuff is not there. I asked questions on the Unity forums with regard to how to package it but to no avail so far.

So I've realised probably the best bet is to port my custom MDX code from C# to Java, fairly straightforward, so I was going to base it on this OLAP4J stuff I found.

It must work against any remote OLAP and not just yours, I happen to be using yours as I have a long history with you all, since 2010 QFS days, haha so my natural choice would be one of your cubes.

**Reply by andrewyang94** [reply]
Hey Leigh,

Can you re-state your question one more time? I conferred with my colleagues but I think we may need a clarification on the exact question that you're looking for.

My understanding is that you're trying to perform a custom implementation and use https://github.com/olap4j/olap4j, but this doesn't seem to be working for you with Atoti. Though the reason you're using https://github.com/olap4j/olap4j is because you want to not only connect to Atoti via your custom application but also other OLAP vendors/applications as well, correct? And that the final follow-up question is if we have any recommendations or toolkits to help solve your issue/unblock you?

**Comment by LeighTilley** [comment]
Hey hey mate

Thanks alot for a reply; I wasn't expecting one! :)

So, here is the situation:

![image](https://github.com/user-attachments/assets/0568edfd-5a8c-43fd-b67a-1a81d7b64ee2)


I've since read that actually fairly recently there is an ADOMD version in .Net Core, so perhaps this will work. Unity packages .Net differently so although it's C# it isn't the same as official .Net / C# versions, with something called '.Net Standard 2.1' and a newer 3 I think. I test function from my Unity C#utilising my MDX DLL works up until trying to connect and then there is an error (basically that the underlying expected core MS libs are not there - usual behaviour for different OS and I'd totally forgotten that Unity has a 'cut-down' .Net packaged with it). With .Net core, more platforms, SSAS on Azure they've released newer libs apparently.

So after a while I thought 'I know, I'll just port my C# code to Java, as I can write in both' (they are almost the same really).  I did the reverse in 2022-23 when CACIB forced me to replace ActivePivot with MS SSAS; I ported my real-time codebase from Java to C# to provide real-time data from MurexAPI. I went looking for the connector code with which to use with my code and I stumbled across (the very old..2013 ish) OLAP4J. Running it against your cube I kept getting this getFlavor error. Once I get the cube connection and query code working I can return to the more funky Unity AR stuff.

Currently I see my options as:

- Go into the old OLAP4J source and edit the getFlavor function and recompile the code myself to then try my query stuff (what worries me is that works initially and then I encounter other weird stuff as it's s old..)
- Search and obtain the .Net core ADOMD libraries and attempt to get this working with my C# code
- Find a newer alternative Java framework to use for connecting (as although some stuff is in C# Unity can use Java libs especially as my target platform is Android on ARMv7 for the AR app)

So yes you are correct, as usual :), in that 'And that the final follow-up question is if we have any recommendations or toolkits to help solve your issue/unblock you?'

Thanks alot

Leigh
https://tilleytech.com

P.S. I'm not sure if you will be able to read this, I found the question, unanswered, on Unity forums from last year:

https://discussions.unity.com/t/my-dll-works-however-ms-adomdclient-dll-i-call-does-not/952381

**Reply by andrewyang94** [reply]
Hmmm... Yeah unfortunately I don't think we can really provide support here in terms of your custom implementation. As you had mentioned, https://github.com/olap4j/olap4j's last commit was 9 years ago, and when I inspect a bit further [here](https://github.com/olap4j/olap4j?tab=readme-ov-file#version-20) it looks like the devs had wanted to enhance compatibility but never got around to it. I suspect that's what you're seeing in your code, where the compatibility of the olap4j library itself is just too outdated and hence why using it to connect to the Atoti cube is giving you hiccups.

Though have you tried the example illustrated on their GitHub README? https://github.com/olap4j/olap4j?tab=readme-ov-file#writing-a-simple-program. I know you're trying to use the XMLA driver but I see you're implementing it differently than what they show in their example. Could be worth a shot, but if it doesn't work with our XMLA endpoint (e.g. `localhost:9090/xmla`) I would chalk it up to compatibility issues of an outdated library. 

Regarding any advice on how to unblock you? Unfortunately, I think that's outside of the scope we can provide support for. We can help support you on how to use Atoti and any issues that arise directly from within the Atoti ecosystem, but we may not be the best resource for triaging other third-party libraries' compatibility with Atoti. Sorry, we hope you can understand and apologies we can't be of more help 🙏

**Comment by LeighTilley** [comment]
Hey hey

Yeah no worries mate and thanks alot for following up as I didn't expect it. :)

This is very much a special op haha and something I've had in my head for ages. I don't think I do need to port the code to Java; I'm convinced there's a way to make a build of my DLL for use with Unity and an Android / ARMv7 deployment and it lies in getting the right combo of target platform and associated APIs. Earlier today I found a table on one of the many lengthy MS webpages with the various .Net and .Net Standard stuff, which is supported where, and in which Android and Unity versions/builds etc.

So I will continue to play around with it until I crack it! :)

Leigh
https://tilleytech.com

**Reply by LeighTilley** [reply]
Hi

Very long time no speak! 

So I made lots of progress with my special project and I realised eventually I could not use my custom C# using MS libraries so I wrote new C# and got XMLA envelopes working instead :). This way I'm not hampered by needing dependencies to work on the ARM64 Android I land on.

I've been making great progress and then today I had to clean and update my Atoti CE to the latest version 0.9.14. I was making some new updates to my special Unity project and then when I ran it I thought it was me....so I had to debug a bit. It seems that something may have changed in 0.9.14 such that my XMLA Discovery commands for Measures and Dimensions does not work anymore.

```
     //Discover commands
     private static string DISCOVER_HIERARCHIES_COMMAND = "MDSCHEMA_HIERARCHIES";
     private static string DISCOVER_MEASURES_COMMAND = "MDSCHEMA_MEASURES";

        private static string DiscoverSoapEnvelope(string discoverCommand)
        {
            return $@"<?xml version=""1.0"" encoding=""utf-8""?>
                    <soap:Envelope 
                    xmlns:soap=""http://schemas.xmlsoap.org/soap/envelope/""
                    xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance""
                    xmlns:xsd=""http://www.w3.org/2001/XMLSchema"">
                    <soap:Body>
                    <Discover xmlns=""urn:schemas-microsoft-com:xml-analysis"">
                    <RequestType>{discoverCommand}</RequestType >
                    <Restrictions>
                        <RestrictionList>
                        <CATALOG_NAME>{catalog}</CATALOG_NAME>
                        <CUBE_NAME>{cube}</CUBE_NAME>
                        </RestrictionList>
                    </Restrictions>
                    <Properties>
                        <PropertyList>
                        <Catalog></Catalog>
                        </PropertyList>
                    </Properties>
                    </Discover>                    
                    </soap:Body>
                 </soap:Envelope>";
        }
```

catalog = atoti
cube = Sales

This has been working perfectly for a long time. My ExcelDNA XLL plugin works although that uses MS ADoMdNet library.

Fortunately I can do wireless debugging to check stuff, so here we can see that I used to get measures back and now I am getting 0:

<img width="1988" height="1312" alt="image" src="https://github.com/user-attachments/assets/59bedb35-6279-4755-af26-d13b30f107ff" />


I thought I'd ask you and I will look at my discovery stuff again, although this has not changed in months since I got it working.

So has something in the way XMLA discovery works changed?

Thanks in advance for your time and assistance! :)

Thanks

Leigh
tilleytech.com

**Reply by LeighTilley** [reply]
Hey no worries - I got Claude to check the envelope and it seems something your side changed and removing my restriction clause on CUBE_NAME in the discovery XMLA query fixed it! :)

Thanks

Leigh
tilleytech.com
